### PR TITLE
Redlist data

### PR DESCRIPTION
**Pull request by Levia**
## Description

This is the first draft to import redlist data from CSVs within the database.
After running migrations using `rake db:migrate` you can import the data using `rake import:redlist_data` which will populate the `species` table and `geo_entities_species` join table.

There have been a few amendments to the files which I will list here briefly:

* files encoding converted to UTF-8 where necessary
* `species.csv` files header refactored to match column names in the db
* `coldwatercorals` folder renamed `coldcorals` to match habitat name in the db
* `coralreef` folder renmaed `colalreefs` to match habitat name in the db
* `country-species-join.csv` file within `coldcorals` has been renamed using only lowercase characters.

Other amendments to do for the future:

* Make sure headers follow a provided template
* Ideally, headers should follow database column names naming (so lowercase with underscores etc.)
* CSV files should not contain empty rows (currently some files do have empty rows)

### CODE REVIEW

File: lib/data/species/coldcorals/citations-utf8.csv [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+20,North-East Atlantic - potentially a lot of data!!!,"Deep-sea coral distribution on seamounts, oceanic islands, and continental slopes in the Northeast Atlantic",J. Hall-Spencer,A. Rogers,et al.,2007,"Conservation and adaptive management of seamount and deep-sea coral ecosystems. Rosenstiel School of Marine and Atmospheric Science, University of Miami",https://www.ingentaconnect.com/content/umrsmas/bullmar/2007/00000081/a00103s1/art00013?crawler=true
+21,Pacific,"Forever in the dark: the cave-dwelling azooxanthellate reef coral Leptoseris troglodyta sp. N. (Scleractinia, Agariciidae)",B. W. Hoeksema,,,2012,Zookeys,https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3487639/
+,,,,,,,,
+,,,,,,,,
```
Comment (drewpotter) [inline lib/data/species/coldcorals/citations-utf8.csv]: I have noticed a lot of blank records from line 23 onwards and wondered if that was meant to be like that?

File: lib/data/species/coldcorals/species-utf8.csv [hunk 2]
```diff
@@ -0,0 +1,2300 @@
...
+909,Isidella longiflora,Isidella,longiflora,NE,N/A,NE,,,,,,
+910,Isidella tentaculum,Isidella,tentaculum,NE,N/A,NE,,,,,,
+911,Isidella tenuis,Isidella,tenuis,NE,N/A,NE,,,,,,
+912,Isidella trichotoma,Isidella,trichotoma,NE,N/A,NE,,,,,,
+913,Janaria mirabilis,Janaria,mirabilis,NE,N/A,NE,,,,,,
+914,Javania antarctica,Javania,antarctica,NE,N/A,NE,,,,,,
+915,Javania borealis,Javania,borealis,NE,N/A,NE,,,,,,
+916,Javania cailleti,Javania,cailleti,NE,N/A,NE,,,,,,
+917,Javania californica,Javania,californica,NE,N/A,NE,,,,,,
+918,Javania cristata,Javania,cristata,NE,N/
```
Comment (drewpotter) [inline lib/data/species/coldcorals/species-utf8.csv]: More blank records again in this CSV

File: lib/data/species/coralreefs/countrylist-utf8.csv [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+Ukraine,UKR,0,1; 4;,0
+United Kingdom of Great Britain and Northern Ireland,GBR,0,1; 4;,0
+Uruguay,URY,1,N/A,0
+Uzbekistan,UZB,0,Inland,0
+Venezuela (Bolivarian Republic of),VEN,2,1;,0
+Vietnam,VNM,2,1;,0
+Zambia,ZMB,0,Inland,0
+Zimbabwe,ZWE,0,Inland,0
+,,,,
+,,,,
```
Comment (drewpotter) [inline lib/data/species/coralreefs/countrylist-utf8.csv]: Again, more blank records

File: lib/data/species/mangroves/citations-utf8.csv [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+23,Papua New Guinea,Masculinity and the culture of rising sea-levels in a mangrove lagoon in Papua New Guinea,D Lipset,,,2017,Maritime studies,https://link.springer.com/article/10.1186/s40152-017-0057-5
+24,West Africa,The assessment of mangrove biomass and carbon in West Africa: a spatially explicit analytical framework,W. Tang,W. Feng,et al.,2016,Wetlands Ecological Management,https://link.springer.com/content/pdf/10.1007%2Fs11273-015-9474-7.pdf
+25,Western Indian Ocean,Genetic population structure of the mangrove snails Littoraria subvittata and L. pallescens in the Western Indian Ocean,A. Nehemia,Y. Ngendu,M. Kochzius,2019,Journal of Experimental Marine Biology and Ecology,http://41.89.141.8/kmfri/bitstream/123456789/1109/1/nehemia2019.pdf
+26,Southeast Asia,"Rates and drivers of mangrove deforestation in Southeast Asia, 2000 - 2012",D. R. Richards,D. A. Friess,,2015,PNAS,https://www.pnas.org/content/pnas/113/2/344.full.pdf
+,,,,,,,,
+,,,,,,,,
```
Comment (drewpotter) [inline lib/data/species/mangroves/citations-utf8.csv]: Again here.

File: lib/tasks/import_redlist_data.rake [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+      import_countries_species(habitat_obj)
+    end
+  end
+
+  def import_species(habitat)
+    species_header = Species.new.attributes.keys.freeze
+    species_filename = "lib/data/species/#{habitat.name}/species-utf8.csv".freeze
+
+    CSV.foreach(species_filename, headers: true) do |row|
+      # TODO Remove empty rows from CSVs
```
Comment (drewpotter) [inline lib/tasks/import_redlist_data.rake]: Ah I see you already spotted this. So my above comments are not necessary!

File: lib/tasks/import_redlist_data.rake [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+      next unless row['species_id']
+      next if Species.find_by(species_id: row['species_id']).present?
+      row_hash = row.to_hash.slice(*species_header)
```
Comment (drewpotter) [inline lib/tasks/import_redlist_data.rake]: Nice! I didn't think of using `slice` for this 🎉

File: lib/tasks/import_redlist_data.rake [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    end
+  end
+
+  def import_countries_species(habitat)
+    countries_species_filename = "lib/data/species/#{habitat.name}/country_species_join-utf8.csv".freeze
+
+    CSV.foreach(countries_species_filename, headers: true) do |row|
+      geo_entity = GeoEntity.find_by(iso3: row['ISO3'])
+      next unless geo_entity.present?
+      GeoEntitiesSpecies.create(species_id: row['SpeciesID'], geo_entity_id: geo_entity.id)
```
Comment (drewpotter) [inline lib/tasks/import_redlist_data.rake]: I am wondering if we should remove the capitalisation from `ISO3` and `SpeciesID`?

**Comment by Levia** [comment]
Thank you @drewpotter . Your comment has been addressed.
I've also amended the CSV files slightly so that they have more consistent headers and filenames. Also removed empty records for a file.

We should share these templates with the team so we are sure we are using the same one.

I think this could be merged in one of your other branches.

**Comment by drewpotter** [comment]
Thanks @Levia I will merge it into the existing `ticket/38/csv-parser-habitat-data` branch if that is okay? I wasn't sure about other branches - please advise 😄

**Comment by Levia** [comment]
@drewpotter I see that your branch has not been merged yet.
Are you waiting for something specific before doing that?

I think that would be the best place where to merge this anyway, as this branch is based on yours.

**Comment by drewpotter** [comment]
@Levia I am trying to add some tests/get some other tests to pass. However, since that is holding things up; I could skip all the tests I am working on temporarily - and merge your branch into mine and request a PR review of mine. And then reopen the testing issues in a fresh branch. Does that sound like a good idea?

**Comment by Levia** [comment]
@drewpotter Yes, let's go for that.
I think that branch is quite a piece of work on its own, so I don't see any benefit in adding more stuff in. 
As you said, you could start a new branch for tests

**Pull request merged**